crypto/pkcs7/pkcs7.c

```diff
@@ -1086,3 +1062,347 @@ OPENSSL_END_ALLOW_DEPRECATED
+  if (!EVP_CipherInit_ex(evp_ctx, cipher, NULL, NULL, NULL, 0)) {
+    goto err;
+  }
+  uint8_t iv[EVP_MAX_IV_LENGTH];
+  OPENSSL_memcpy(iv, enc_alg->parameter->value.octet_string->data,
+                 enc_alg->parameter->value.octet_string->length);
+  const int expected_iv_len = EVP_CIPHER_CTX_iv_length(evp_ctx);
+  if (enc_alg->parameter->value.octet_string->length != expected_iv_len) {
+    OPENSSL_PUT_ERROR(PKCS7, ERR_R_PKCS7_LIB);
+    goto err;
+  }
+  if (!EVP_CipherInit_ex(evp_ctx, NULL, NULL, NULL, iv, 0)) {
+    goto err;
+  }
+  // Get the key length from cipher context so we don't condition on |cek_len|
+  int len = EVP_CIPHER_CTX_key_length(evp_ctx);
+  if (!len) {
+    goto err;
+  }
+  // Always generate random bytes for the dummy key, regardless of |cek| decrypt
+  dummy_key = OPENSSL_malloc(len);
+  RAND_bytes(dummy_key, len);
+  // At this point, null |cek| indicates that no content encryption key was
+  // successfully decrypted. We don't want to return early due to MMA. So, swap
+  // in the dummy key and proceed. Content decryption result will be gibberish.
+  if (cek == NULL) {
+    cek = dummy_key;
+    dummy_key = NULL;
+  }
+
+  if (!EVP_CipherInit_ex(evp_ctx, NULL, NULL, cek, NULL, 0)) {
+    goto err;
+  }
+
+  OPENSSL_free(cek);
+  OPENSSL_free(dummy_key);
+  out = cipher_bio;
+
+  if (data_body && data_body->length > 0) {
+    data_bio = BIO_new_mem_buf(data_body->data, data_body->length);
+  } else {
+    data_bio = BIO_new(BIO_s_mem());
+    if (data_bio == NULL) {
+      goto err;
+    }
+    BIO_set_mem_eof_return(data_bio, 0);
+  }
+  if (data_bio == NULL) {
+    goto err;
+  }
+  BIO_push(out, data_bio);
+  return out;
+
+err:
+  OPENSSL_free(cek);
+  OPENSSL_free(dummy_key);
+  BIO_free_all(out);
+  BIO_free_all(cipher_bio);
+  BIO_free_all(data_bio);
+  return NULL;
+}
+
+PKCS7_RECIP_INFO *PKCS7_add_recipient(PKCS7 *p7, X509 *x509) {
+  PKCS7_RECIP_INFO *ri;
+  if ((ri = PKCS7_RECIP_INFO_new()) == NULL ||
+      !PKCS7_RECIP_INFO_set(ri, x509) || !PKCS7_add_recipient_info(p7, ri)) {
+    PKCS7_RECIP_INFO_free(ri);
+    return NULL;
+  }
+  return ri;
+}
+
+int PKCS7_decrypt(PKCS7 *p7, EVP_PKEY *pkey, X509 *cert, BIO *data, int flags) {
+  GUARD_PTR(p7);
+  GUARD_PTR(pkey);
+  GUARD_PTR(data);
+  BIO *bio = NULL;
+  int ret = 0;
+
+  switch (OBJ_obj2nid(p7->type)) {
+    case NID_pkcs7_enveloped:
+    case NID_pkcs7_signedAndEnveloped:
+      break;
+    default:
+      OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_WRONG_CONTENT_TYPE);
+      goto err;
+  }
+
+  if (cert && !X509_check_private_key(cert, pkey)) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_PRIVATE_KEY_DOES_NOT_MATCH_CERTIFICATE);
+    goto err;
+  }
+
+  if ((bio = pkcs7_data_decode(p7, pkey, cert)) == NULL ||
+      !pkcs7_bio_copy_content(bio, data)) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_DECRYPT_ERROR);
+    goto err;
+  }
+
+  // Check whether content decryption was successful
+  OPENSSL_BEGIN_ALLOW_DEPRECATED
+  if (!BIO_get_cipher_status(bio)) {
```

Comment:
`BIO_get_cipher_status` could return `-2`, so it's probably best we make this check more specific

Reply:
maybe for openssl, but I don't think it can return -2 for us. we're pretty careful in [the cipher BIO](https://github.com/aws/aws-lc/blob/main/crypto/pkcs7/bio/cipher.c) to keep `ctx->ok` either 0 or 1.

Reply:
Yeah I noticed that we were doing everything right in `cipher bio`'s code, it's just [this place](https://github.com/aws/aws-lc/blob/80f984ef78203bbdb61a2ef3fee55d8956ad7e8c/crypto/bio/bio.c#L361-L364) in `BIO_get_cipher_status` that leaves me wary.
It's unlikely, but if any user passes in a `bio` with weird function pointers, it may trigger a success here?

Reply:
oh, yeah. i didn't notice that in `BIO_ctrl`. we should handle that correctly here. will update.